# **Association Rules**

Association Rules Analysis has become familiar for analysis in the retail industry. It is also called Market Basket Analysis terms. This analysis is also used for advice. Personal recommendations in applications such as Spotify, Netflix, and Youtube can be given as examples. 
[](http://)
One of the biggest examples of Association Rules Analysis is the correlation between beer and diaper. When Walmart, a chain store in the United States, studied the shopping behavior of customers, the study showed that diapers and beers are bought together. Because, as it turns out, fathers are often tasked with shopping while mothers stay with the baby.


# **Apriori Algorithm**

The Apriori Algorithm, used for the first phase of the Association Rules, is the most popular and classical algorithm in the frequent old parts. These algorithm properties and data are evaluated with Boolean Association Rules. In this algorithm, there are product clusters that pass frequently, and then strong relationships between these products and other products are sought.

The importance of an Association Rules can be determined by 3 parameters that are used to identify the strength of the algorithm. Namely,

* Support
* Confidence
* Lift


Let X and Y represent the products in the market and N represents the total number of products.

**Support**     :  It is the probability of an event to occur.

**Confidence**  :  It is a measure of conditional probability

**Lift**        : It is the probability of all items occurring together divided by the product of antecedent and consequent occurring as if they are independent of each other.



In [1]:
#Importing the libraries and the dataset
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [2]:
df = pd.read_csv('G:\codeclause\Market Basket Analysis\GroceryStoreDataSet.csv', names=['products'], sep=',')

<>:1: DeprecationWarning: invalid escape sequence \c
<>:1: DeprecationWarning: invalid escape sequence \c
<>:1: DeprecationWarning: invalid escape sequence \c
C:\Users\DELL GUNA\AppData\Local\Temp\ipykernel_1828\1595519703.py:1: DeprecationWarning: invalid escape sequence \c
  df = pd.read_csv('G:\codeclause\Market Basket Analysis\GroceryStoreDataSet.csv', names=['products'], sep=',')


In [3]:
df.head()

products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [4]:
#Checking the dimensions of dataset
df.shape

(20, 1)

In [5]:
#Let's spilt the products and create a list called by 'data'
data = list(df['products'].apply(lambda x: x.split(",")))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

## **Apriori Algorithm and One-Hot Encoding**
 
Apriori's algorithm transforms True/False or 1/0.  
Using TransactionEncoder, we convert the list to a One-Hot Encoded Boolean list.  
Products that customers bought or did not buy during shopping will now be represented by values 1 and 0.

In [6]:
#Transforming the list, with one-hot encoding
from mlxtend.preprocessing import TransactionEncoder
a = TransactionEncoder()
a_data = a.fit(data).transform(data)
df = pd.DataFrame(a_data, columns=a.columns_)
df = df.replace(False, 0)
df

BISCUIT  BOURNVITA  BREAD  COCK  COFFEE  CORNFLAKES  JAM  MAGGI  MILK  \
0       1.0        0.0    1.0   0.0     0.0         0.0  0.0    0.0   1.0   
1       1.0        0.0    1.0   0.0     0.0         1.0  0.0    0.0   1.0   
2       0.0        1.0    1.0   0.0     0.0         0.0  0.0    0.0   0.0   
3       0.0        0.0    1.0   0.0     0.0         0.0  1.0    1.0   1.0   
4       1.0        0.0    0.0   0.0     0.0         0.0  0.0    1.0   0.0   
5       0.0        1.0    1.0   0.0     0.0         0.0  0.0    0.0   0.0   
6       0.0        0.0    0.0   0.0     0.0         1.0  0.0    1.0   0.0   
7       1.0        0.0    1.0   0.0     0.0         0.0  0.0    1.0   0.0   
8       0.0        0.0    1.0   0.0     0.0         0.0  1.0    1.0   0.0   
9       0.0        0.0    1.0   0.0     0.0         0.0  0.0    0.0   1.0   
10      1.0        0.0    0.0   1.0     1.0         1.0  0.0    0.0   0.0   
11      1.0        0.0    0.0   1.0     1.0         1.0  0.0    0.0   0.0   
12      0.0        1.0    0.0   0.0     1.0         0.0  0.0    0.0   0.0   
13      0.0        0.0    1.0   1.0     1.0         0.0  0.0    0.0   0.0   
14      1.0        0.0    1.0   0.0     0.0         0.0  0.0    0.0   0.0   
15      0.0        0.0    0.0   0.0     1.0         1.0  0.0    0.0   0.0   
16      0.0        1.0    1.0   0.0     0.0         0.0  0.0    0.0   0.0   
17      0.0        0.0    1.0   0.0     1.0         0.0  0.0    0.0   0.0   
18      0.0        0.0    1.0   0.0     1.0         0.0  0.0    0.0   0.0   
19      0.0        0.0    0.0   0.0     1.0         1.0  0.0    0.0   1.0   

    SUGER  TEA  
0     0.0  0.0  
1     0.0  0.0  
2     0.0  1.0  
3     0.0  0.0  
4     0.0  1.0  
5     0.0  1.0  
6     0.0  1.0  
7     0.0  1.0  
8     0.0  1.0  
9     0.0  0.0  
10    0.0  0.0  
11    0.0  0.0  
12    1.0  0.0  
13    0.0  0.0  
14    1.0  0.0  
15    1.0  0.0  
16    1.0  0.0  
17    1.0  0.0  
18    1.0  0.0  
19    0.0  1.0

## **Applying Apriori and Resulting**

The next step is to create the Apriori Model. We can change all the parameters in the Apriori Model in the mlxtend package.   
I will try to use minimum support parameters for this modeling.    
For this, I set a min_support value with a threshold value of 20% and printed them on the screen as well.

In [7]:
#Setting a threshold value for the support and calculate the support value.
df = apriori(df, min_support=0.2, use_colnames=True, verbose=1)
df

Processing 42 combinations | Sampling itemset size 3


c:\users\dell guna\appdata\local\programs\python\python37\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:115: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  DeprecationWarning,


support              itemsets
0      0.35             (BISCUIT)
1      0.20           (BOURNVITA)
2      0.65               (BREAD)
3      0.40              (COFFEE)
4      0.30          (CORNFLAKES)
5      0.25               (MAGGI)
6      0.25                (MILK)
7      0.30               (SUGER)
8      0.35                 (TEA)
9      0.20      (BISCUIT, BREAD)
10     0.20         (MILK, BREAD)
11     0.20        (SUGER, BREAD)
12     0.20          (TEA, BREAD)
13     0.20  (CORNFLAKES, COFFEE)
14     0.20       (SUGER, COFFEE)
15     0.20          (TEA, MAGGI)

Choosing 60% minimum confidance value. In other words, when product X is purchased, we can say thst the purchase of product Y is 60% or more.

In [8]:
#Let's review the interpretation values using the Association rule function.
df_ar = association_rules(df, metric="confidence", min_threshold=0.6)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0        (MILK)     (BREAD)                0.25                0.65      0.2   
1       (SUGER)     (BREAD)                0.30                0.65      0.2   
2  (CORNFLAKES)    (COFFEE)                0.30                0.40      0.2   
3       (SUGER)    (COFFEE)                0.30                0.40      0.2   
4       (MAGGI)       (TEA)                0.25                0.35      0.2   

   confidence      lift  leverage  conviction  
0    0.800000  1.230769    0.0375        1.75  
1    0.666667  1.025641    0.0050        1.05  
2    0.666667  1.666667    0.0800        1.80  
3    0.666667  1.666667    0.0800        1.80  
4    0.800000  2.285714    0.1125        3.25

## Inferences
For example, if we examine our 0th index value:
- The probability of seeing milk sales is seen as 25%.
- Bread intake is seen as 65%.
- We can say that the support of both of them is measured as 20%.
- 80% of those who buys bread, buys bread as well.
- Their correlation with each other is seen as 1.75.

As a result, if item X and Y are bought together more frequently, then several steps acn be taken to increase the profit. For instance:
- Cross-Selling can be improved by combining products - items.
- The shop layout can be changed so that the sales can be improved when certain items are kept together.
- Promotional activities which are an advertising campaign can be carried out to increase the ales of goods that customers do not buy.
- Collective discounts can be offered on these products id the customer buys both of them.